lib/phoenix_live_view/diff.ex

```diff
@@ -84,6 +84,37 @@ defmodule Phoenix.LiveView.Diff do
     end
   end
 
+  @doc """
+  Updates the components with a list of `send_update` operations.
+
+  Like `with_component/5`, it will store the result under the `cid
+   key in the `component_diffs` map.
+
+  It returns the updated `component_diffs` and the updated `components` or
+  raises an `ArgumentError` if no component matching the update `:id` exists.
+
+  Components are preloaded before the update callback.
+
+  ## Example
+
+    {diffs, new_components} = Diff.update_components(socket, state.components, updates)
+  """
+  def update_components(socket, components, updates) do
```

Comment:
Given the comment above and the fact we are not doing any grouping here (i.e. preload is still called once per component instance), my suggestion is to make this function work exclusively on a single component.